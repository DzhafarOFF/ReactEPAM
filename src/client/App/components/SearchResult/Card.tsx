import './Card.scss';
import React, { useCallback } from 'react';
import { Link } from 'react-router-dom';
import { Movie } from '../../typings/types';
import { setCurrentMovie } from '../../actions/setCurrentMovie';
import { useDispatch } from 'react-redux';

const Card: React.FC<Movie> = (props: Movie) => {
	const getGenresFromArray = (genres: string[]): string => genres.reduce((genre, resultString) => {
		const firstLetter = genre.charAt(0).toUpperCase();
		const restLettersFrom = 1;
		const restLetters = genre.slice(restLettersFrom);
		const capitalizeGenre = firstLetter + restLetters;
		return `${resultString} ${capitalizeGenre}`;
	}, '');
	const dateStringLength = 4;
	const date = props.releaseDate.substring(0, dateStringLength);

	const dispatch = useDispatch();
	const selectCard = useCallback(() => dispatch(setCurrentMovie(props)),[dispatch, props]);

	return (
		<div className = 'result-card' onClick = {selectCard}>
			<Link to={`/films/${props.id}`} className='link'>
				<img className = 'card-image' src = {props.imageURL}/>
				<div className = 'card-content'>
					<h3 className = 'card-title'>{props.title}</h3>
					<p className = 'card-date'>{date}</p>
				</div>
				<p className = 'card-genres'>{getGenresFromArray(props.genres)}</p>
			</Link>
		</div>
	);
};

export default Card;
