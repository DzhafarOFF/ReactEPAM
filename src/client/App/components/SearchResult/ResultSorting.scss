.sorting-container {
    background-color: #F5F5F5;
}

.sorting{
    display: flex;
    justify-content: space-around;
    align-items: center;
    color: #131313;
}

.sorting-title {
    display: inline-block;
}

.sorting-counter {
    margin: 0;
    display: inline-block;
}