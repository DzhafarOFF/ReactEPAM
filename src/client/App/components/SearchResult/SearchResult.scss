.search-result{
  display: grid;
  grid-template-columns: repeat(auto-fill, 300px);
  grid-gap: 20px;
  justify-content: space-between;
}
.search-result-container{
    max-width: 1140px;
    margin: 0 auto;
    margin-top: 20px;
}