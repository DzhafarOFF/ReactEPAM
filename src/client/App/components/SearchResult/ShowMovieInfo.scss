.movie-modal-container {
    background-image: url(../Search/netflixroulette.jpg);
    background-size: cover;
    background-color: #404040;
    background-blend-mode: overlay;
}
.movie-modal-header{
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.movie-modal{
    display: flex;
    flex-direction: column;
    margin: 0 auto;
    padding: 40px;
    max-width: 1140px;
}
.movie-content-container{
    display: flex;
    justify-content: space-between;
    align-items: center;
}
.movie-modal-info {
    text-transform: none;
    max-width: 600px;
    font-size: 1.5rem;
    color: #fefefe;
}
.movie-image {
    width: 300px;
}
.movie-modal-title{
    font-size: 2rem;
    text-transform: uppercase;
    color: lighten($color: #E50914, $amount: 20);
}
.movie-rating {
    border: 2px solid;
    display: inline-block;
    padding: 10px 5px;
    border-radius: 50%;
}

.movie-genres{
    margin-top: 0;
}
.movie-release,
.movie-runtime{
    font-weight: bolder;
}