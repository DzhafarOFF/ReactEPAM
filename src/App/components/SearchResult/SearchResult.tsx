import React from 'react';
import Card from './Card';
import { MovieProps } from './types';

type Props = {
    movies: MovieProps[];
}

const SearchResult = (props: Props) => {
    const setSearchResult = props.movies.map(
        (movie: MovieProps) => <Card
            key = {movie.key}
            title = {movie.title}
            genres = {movie.genres}
            releaseDate = {movie.releaseDate}
            imageURL = {movie.imageURL}
        />
    );
    return (
        <div>
            {setSearchResult}
        </div>
    );
}

export default SearchResult;