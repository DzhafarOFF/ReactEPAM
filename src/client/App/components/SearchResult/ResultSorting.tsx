import './ResultSorting.scss';
import Button from '../common/Button';
import React from 'react';

type Props = {
	numberOfItems: number
	sortingOptions: string[]
	activeOption: string
	handleSwitchSorting: (e: React.MouseEvent) => void
};

const ResultSorting: React.FC<Props> = (props: Props) => {
	const setSortingOption = props.sortingOptions.map((option, index) =>
		<Button
			content = {option}
			key = {index}
			action = {props.handleSwitchSorting}
		/>
	);
	return (
		<div className = 'sorting-container'>
			<div className = 'sorting'>
				<p className = 'sorting-counter'>{props.numberOfItems} {'movies found'}</p>
				<div>
					<h3 className = 'sorting-title'>{'Sort by'}</h3>
					{setSortingOption}
				</div>
			</div>
		</div>
	);
};

export default ResultSorting;
