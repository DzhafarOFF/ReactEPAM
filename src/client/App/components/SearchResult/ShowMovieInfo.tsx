import './ShowMovieInfo.scss';
import React, { useCallback } from 'react';
import { Link } from 'react-router-dom';
import { Movie } from '../../typings/types';
import { removeCurrentMovie } from '../../actions/setCurrentMovie';
import { useDispatch } from 'react-redux';

const ShowMovieInfo: React.FC<Movie> = (props: Movie) => {
	const dispatch = useDispatch();
	const selectSearch = useCallback((): void => {
		dispatch(removeCurrentMovie());
	}, [dispatch]);

	const dateStringLength = 4;
	const date = props.releaseDate.substring(0, dateStringLength);

	return (
		<div className = 'movie-modal-container'>
			<div className = 'movie-modal'>
				<div className = 'movie-modal-header'>
					<Link to='/' className = 'search-title'>
						<h1>{'Netflixroulette'}</h1>
					</Link>
					<Link to='/'>
						<button className = 'button' onClick = {selectSearch}>{'Search'}</button>
					</Link>
				</div>
				<div className = 'movie-content-container'>
					<img className = 'movie-image' src = {props.imageURL}/>
					<div className = 'movie-modal-info'>
						<div className = 'movie-content-container'>
							<h3 className = 'movie-modal-title'>{props.title}</h3>
							<p className = 'movie-rating'>{props.rating}</p>
						</div>
						<p className = 'movie-genres'>{props.genres.join(' ')}</p>
						<div className = 'movie-content-container'>
							<p className = 'movie-release'>{date}</p>
							<p className = 'movie-runtime'>{props.runtime} {'min'}</p>
						</div>
						<p>{props.description}</p>
					</div>
				</div>
			</div>
		</div>
	);
};

export default ShowMovieInfo;
