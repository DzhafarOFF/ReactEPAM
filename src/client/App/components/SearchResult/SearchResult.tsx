import React from 'react';
import Card from './Card';
import ResultSorting from './ResultSorting';
import { Movie} from './types';
import './SearchResult.scss';

type Props = {
    action: (movie: Movie) => void;
    movies: Movie[];
}

type State = {
    sortBy: {
        [key: string]: boolean;
    };
    activeSortingOption: string;
}
class SearchResult extends React.Component<Props, State>{
    constructor(props: Props) {
        super(props)
        this.state = {
            sortBy: {
                releaseDate: true,
                rating: false,
            },
            activeSortingOption: 'releaseDate',
        }
    }

    sortByReleaseDate = (moviesArray: Movie[]) => {
        return moviesArray.sort((first: Movie, second: Movie) => first.releaseDate
            .localeCompare(second.releaseDate));
    }

    sortByRating = (moviesArray: Movie[]) => {
        return moviesArray.sort((first: Movie, second: Movie) => first.rating - second.rating);
    }

    switchSorting = (e: React.MouseEvent) => {
        const newOption = (e.target as HTMLInputElement).value;
        const active = Object.keys(this.state.sortBy).find(key => this.state.sortBy[key]);
        const newState = {
            sortBy: {
                ...this.state.sortBy, 
                [newOption]: true,
                [active]: false
            },
            activeSortingOption: newOption
        }
        if (newOption !== active) this.setState(newState); 
    }

    handleClick = (movie: Movie) => {
        this.props.action(movie);
    }

    render() {
        const moviesSorted = this.props.movies;
        if (this.state.activeSortingOption === 'releaseDate') {
            this.sortByReleaseDate(moviesSorted);
        }
        if (this.state.activeSortingOption === 'rating') {
            this.sortByRating(moviesSorted);
        }
            
        const setSearchResult = moviesSorted.map(
            (movie: Movie, index) =>
                <Card
                    action = {(movie) => this.handleClick(movie)}
                    id = {movie.key}
                    key = {movie.key}
                    title = {movie.title}
                    genres = {movie.genres}
                    releaseDate = {movie.releaseDate}
                    imageURL = {movie.imageURL}
                    rating = {movie.rating}
                    description = {movie.description}
                    runtime = {movie.runtime}
                />
        );
        const sortingOptions = Object.keys(this.state.sortBy);
        return (
            <>
                <ResultSorting
                    numberOfItems = {this.props.movies.length}
                    sortingOptions = {sortingOptions}
                    activeOption = {sortingOptions[0]}
                    handleSwitchSorting = {this.switchSorting}
                />
                <div className = 'search-result-container'>
                    <div className = 'search-result'>
                        {setSearchResult}
                    </div>
                </div>
            </>
        );
    }
}

export default SearchResult;