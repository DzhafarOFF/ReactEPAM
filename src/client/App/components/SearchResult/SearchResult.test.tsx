import React from 'react';
import { shallow, render, mount } from 'enzyme';
import SearchResult from './SearchResult';
const movie = {
    key: 1,
    title: 'MovieName',
    genres: ['Action','Drama'],
    releaseDate: '2018',
    imageURL: 'URL',
    rating: 9,
    description: 'Movie description',
    runtime: 120
}
const props = {
    action: () => {console.log(movie)},
    movies: [
        {
            key: 1,
            title: 'MovieName',
            genres: ['Action','Drama'],
            releaseDate: '2018',
            imageURL: 'URL',
            rating: 9,
            description: 'Movie description',
            runtime: 120
        },
        {
            key: 1,
            title: 'MovieName',
            genres: ['Action','Drama'],
            releaseDate: '2018',
            imageURL: 'URL',
            rating: 9,
            description: 'Movie description',
            runtime: 120
        }
    ]
}
describe('<SearchResult/> rendering', () => {
    const searchResult = shallow(<SearchResult {...props}/>)
    it('should renders correctly',() => {
        expect(searchResult).toMatchSnapshot();
    })
});