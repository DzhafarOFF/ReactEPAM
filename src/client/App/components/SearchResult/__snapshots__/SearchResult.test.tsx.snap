// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResult/> rendering should renders correctly 1`] = `
<Fragment>
  <ResultSorting
    activeOption="releaseDate"
    handleSwitchSorting={[Function]}
    numberOfItems={2}
    sortingOptions={
      Array [
        "releaseDate",
        "rating",
      ]
    }
  />
  <div
    className="search-result-container"
  >
    <div
      className="search-result"
    >
      <Card
        action={[Function]}
        description="Movie description"
        genres={
          Array [
            "Action",
            "Drama",
          ]
        }
        imageURL="URL"
        key="1"
        rating={9}
        releaseDate="2018"
        runtime={120}
        title="MovieName"
      />
      <Card
        action={[Function]}
        description="Movie description"
        genres={
          Array [
            "Action",
            "Drama",
          ]
        }
        imageURL="URL"
        key="1"
        rating={9}
        releaseDate="2018"
        runtime={120}
        title="MovieName"
      />
    </div>
  </div>
</Fragment>
`;
